### PR TITLE
Make it obvious to see all the logs

### PR DESCRIPTION
**Pull request by aanorbel**
Fixes  https://github.com/ooni/probe/issues/1777

## Proposed Changes

  - Include log file appender
  - Fetch and display logs to UI.
  - Add log file sharing.

| . | . | . | . |
|-|-|-|-|
| Screenshot_20220824_185707 | Screenshot_20220824_185741 | Screenshot_20220824_185812 | Screenshot_20220824_185851 |

### CODE REVIEW

**Comment by bassosimone** [comment]
This is fantastic! Thank you! I will review ASAP!

**Comment by bassosimone** [comment]
Feature request: possibility of copying the logs! (We're starting to test this!)

**Comment by bassosimone** [comment]
We discussed today with @aanorbel that we'll implement sharing of the whole logfile via Telegram, etc.

**Comment by bassosimone** [comment]
I have used and testing this functionality a bit. Here's some extra feedback:

- [ ] I see some utility in sharing logs as HTML, but I would rather share them as a text file without tags because I think ultimately the presence of tags would only make log processing more complex on our end

- [ ] the "advanced" menu entry now seems to contain two "debug logs" entries, however I think the one that leads to the logs should not be named "debug logs" but perhaps something like "see recent logs"

I tested the sharing feature and it's working as intended, thank you!

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -166,6 +178,18 @@
 			android:name=".common.service.RunTestJobService"
 			android:label="Run Test Service"
 			android:permission="android.permission.BIND_JOB_SERVICE"/>
+
+		<provider
+			android:name="androidx.core.content.FileProvider"
+			android:authorities="org.openobservatory.ooniprobe.provider"
+			android:exported="false"
+			android:grantUriPermissions="true">
+			<!-- ressource file to create -->
```
Comment (bassosimone) [inline app/src/main/AndroidManifest.xml:187]:
```suggestion
			<!-- resource file to create -->
```

File: app/src/main/java/org/openobservatory/ooniprobe/activity/LogActivity.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import android.view.MenuItem;
+import android.view.View;
+import android.widget.AdapterView;
+import android.widget.ArrayAdapter;
+
+import androidx.core.content.FileProvider;
+import androidx.recyclerview.widget.LinearLayoutManager;
+
+import com.google.common.collect.Lists;
+import com.velmurugan.inapplogger.LogType;
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/LogActivity.java:18]: I am confused by the fact that the `app/build.gradle` imports `com.github.aanorbel`'s `inapplogger` while here we seem to be importing the original implementation. Is this intended?

File: app/src/main/java/org/openobservatory/ooniprobe/activity/LogActivity.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            intentShareFile.setType(URLConnection.guessContentTypeFromName(file.getName()));
+            intentShareFile.putExtra(Intent.EXTRA_STREAM, FileProvider.getUriForFile(
+                            this,
+                            "org.openobservatory.ooniprobe.provider", //(use your app signature + ".provider" )
+                            file)
+                    /*Uri.parse("file://" + file.getAbsolutePath())*/);
+
+            //if you need
+            //intentShareFile.putExtra(Intent.EXTRA_SUBJECT,"Sharing File Subject);
+            //intentShareFile.putExtra(Intent.EXTRA_TEXT, "Sharing File Description");
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/activity/LogActivity.java:91]:
```suggestion
```

Let's remember to zap the commented out code before merging

File: app/src/main/java/org/openobservatory/ooniprobe/common/AppLogger.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
+package org.openobservatory.ooniprobe.common;
+
+import android.content.Context;
+
+import androidx.annotation.NonNull;
+
+import com.velmurugan.inapplogger.InAppLogger;
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/common/AppLogger.java:7]: Same question as above

File: app/src/main/java/org/openobservatory/ooniprobe/common/AppLogger.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    public AppLogger(@NonNull Context context) {
+        logger = new InAppLogger(context);
+        /*TODO(aanorbel): 01/06/2022  Check log retention policy and clear logs appropriately.
+            logs only exist for an app session. Log file grows with each test run making
+            time to load logs to increase.*/
+        logger.deleteOldLog();
+    }
+
+    public void e(String tag, String message) {
+        logger.e(String.format("%s : %s", tag, message));
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/common/AppLogger.java:31]:
```suggestion
        logger.e(String.format("%s: %s", tag, message));
```

File: app/src/main/java/org/openobservatory/ooniprobe/common/AppLogger.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public void w(String tag, String message) {
+        logger.w(String.format("%s : %s", tag, message));
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/common/AppLogger.java:35]:
```suggestion
        logger.w(String.format("%s: %s", tag, message));
```

File: app/src/main/java/org/openobservatory/ooniprobe/common/AppLogger.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public void i(String tag, String message) {
+        logger.i(String.format("%s : %s", tag, message));
```
Comment (bassosimone) [inline app/src/main/java/org/openobservatory/ooniprobe/common/AppLogger.java:39]:
```suggestion
        logger.i(String.format("%s: %s", tag, message));
```

**Review by bassosimone — COMMENTED** [review]
Thanks you! I have used the code and already provided some feedback out of that. This interim review adds a couple more questions and suggestions!

File: app/src/main/java/org/openobservatory/ooniprobe/activity/LogActivity.java (hunk 2 above)
Comment (aanorbel) [reply app/src/main/java/org/openobservatory/ooniprobe/activity/LogActivity.java:18]: i added functionality to the package without changing the package name. 
i had to publish the update under a different package name as the owner has not approved new changes to the package.

Comment (bassosimone) [reply app/src/main/java/org/openobservatory/ooniprobe/activity/LogActivity.java:20]: ✔️

**Comment by bassosimone** [comment]
I am using the diff and I am happy with it. However, I think it would be better if we only selected the "LOG" events and just emitted their severity and content. Currently we're emitting all MK events, which I think only adds confusion.

**Comment by aanorbel** [comment]
We can always change the type of events we log. Establishing the necessary framework for logging and sharing logs was top priority. I would be removing logs for `MK_EVENTS`.

**Comment by bassosimone** [comment]
> We can always change the type of events we log. Establishing the necessary framework for logging and sharing logs was top priority. I would be removing logs for `MK_EVENTS`.

What I suggest is to only log the `MK_EVENTS` that contain log messages. You can identify them by their key which is `log`. There should be a place in the codebase where we already trap this event for logging. That should suffice.

Thank you! 🙌

**Review by bassosimone — APPROVED** [review]
🐳

**Pull request merged**